packages/flutter/lib/src/painting/decoration_image.dart

```diff
@@ -431,6 +435,29 @@ void paintImage({
     // as we apply a nine-patch stretch.
     assert(sourceSize == inputSize, 'centerSlice was used with a BoxFit that does not guarantee that the image is fully visible.');
   }
+
+  // Output size is fully calculated.
+  assert(() {
+    if (debugImageOverheadAllowedInKilobytes != null) {
+      final int acutalSizeKilobytes = (image.width * image.height * 4 * (4/3)) ~/ 1024;
+      final int possibleSizeKilobytes = (outputSize.width * outputSize.height * 4 * (4/3)) ~/ 1024;
+      final int overhead = acutalSizeKilobytes - possibleSizeKilobytes;
+      if (overhead > debugImageOverheadAllowedInKilobytes) {
+        FlutterError.reportError(FlutterErrorDetails(
```

Comment:
reporting these as a full error seems harmful as this is much more of a warning than an error.
It could be reasonable to add a `FlutterError.reportWarning` to handle this and other cases where we want to add warnings to the framework. 

For example, existing errors all indicate that something has gone wrong and the user needs to take action. This just indicates that something might be inefficient. Existing heuristics like truncating all errors after the first error to de-emphasize errors that are follow on failures from other errors would not make sense for cases like this. If you had this error and then a real error then you would want that other error to be emphasized rather than this one which is just a memory usage warning.

Reply:
This is opt-in - the idea is that if the developer has opted in to seeing it, they want to take action if it happens. 

If we had a facility for reporting warnings, that would work fine, but I'd rather not hold this up in trying to develop one and wiring it up in all the right places.

In my terminal, subsequent errors still have enough relevant detail, although it's good they're truncated a bit since you can run into this error on every repaint of the image, which may happen multiple times at the same size. For example, in testing this in a macOS application, I get this over and over again as the mouse moves across the application.

Reply:
I think this is a very useful and actionable warning/error, and I'd love it to be opt-out instead of opt-in if I were developing a Flutter app. (Figuring out what opt-in features to turn on could be a big cognitive cost.)

Reply:
I think we could make it opt-out at some point, but we should start this way so that people have time to fix things without needing a manual roll internally.

Reply:
Being able to make this opt-in by default would be a great goal. Adding an environment variable like
`flutter.imageSizeWarnings` like the `flutter.inspector.structuredErrors` one @helin24 added here could be a good way to achieve that.
https://github.com/flutter/flutter/commit/479f370379cc721d399cb8e3c0bcac79d58d17b4#diff-f01fcfe0b7237b22d168496f69be49baR947

Having the environment variable in addition to a service extension is useful as a service extension cannot be set immediately on app startup which means you miss errors that occur before the first frame renders.